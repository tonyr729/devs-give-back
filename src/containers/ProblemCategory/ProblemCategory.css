.categories-instructions {
  font-size: 4rem;
  padding-top: 2rem;
  color: #f8f8ff;
  line-height: 4rem;
  margin-bottom: .6rem; }

.categories {
  display: block;
  background-color: #607d8b;
  margin: 2rem auto;
  width: 60%;
  height: 22rem;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between; }

.category-button {
  border-radius: 1rem;
  background-color: #03a9f4;
  color: #f8f8ff;
  border: none;
  height: 2rem;
  padding: 0 1rem;
  font-size: 1rem;
  margin: .4rem; }
  .category-button:hover {
    background-color: #005983; }
  .category-button:active {
    background-color: #00364e; }

.selected {
  border-radius: 1rem;
  background-color: #494949;
  color: #8a8a8a;
  border: none;
  height: 2rem;
  padding: 0 1rem;
  font-size: 1rem;
  margin: .4rem; }
