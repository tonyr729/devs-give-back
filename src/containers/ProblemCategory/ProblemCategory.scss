.categories-instructions {
  font-size: 4rem;
  padding-top: 4rem;
  color: #f8f8ff;
  line-height: 4rem;
  margin-bottom: .6rem;
}

.categories {
  display: block;
  background-color: #607d8b;
  margin: 0 auto;
  width: 60%;
  height: 22rem;
}

.category-button {
  border-radius: 1rem;
  background-color: #03a9f4;
  color: #f8f8ff;
  border: none;
  height: 2rem;
  padding: 0 1rem;
  font-size: 1rem;
  margin: .4rem;
}
