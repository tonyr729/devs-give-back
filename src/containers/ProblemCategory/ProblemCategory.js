import React, { Component } from 'react';
import './ProblemCategory.css';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import DatabaseHelper from '../../helpers/DatabaseHelper';
import { addCategories, selectCategories } from '../../actions/actions';
import CategoryButton from '../../components/Statefull/CategoryButton/CategoryButton';
import rightArrow from '../../images/right_arrow.svg';
import leftArrow from '../../images/left_arrow.svg';


export class ProblemCategory extends Component {
  constructor() {
    super();
    this.state = {
      categories: []
    };
    this.database = new DatabaseHelper();
  }


  async componentDidMount() {
    const categories = await this.database.pullCategoriesFromDatabase();
    this.props.addCategories(categories);
  }

  componentWillReceiveProps() {}

  displayCategories = () => {
    let display = null;
    if (this.props.categories) {
      display = this.props.categories.map((category, index) => ( 
        <CategoryButton key={index} selectCategory={this.selectCategory} category={category} />
      ));
    }
    return display;
  }

  selectCategory= (category) => {
    let match = this.state.categories.includes(category);
    if (!match) {
      this.setState({
        categories: [...this.state.categories, category]
      });
    } else {
      let newState = this.state.categories.filter(currCategory => currCategory !== category);
      this.setState({
        categories: newState
      });
    }
  }

  handleSubmit = () => {
    const {title, body, clientID} = this.props;
    this.props.selectCategories(this.state.categories);
    this.database.writeProblemToDatabase(title, body, this.state.categories, clientID);
    this.props.history.push("/problem-created");
  }

  logInCheck = (client) => {
    const value = Object.keys(client).length;
    if ( value === 0 ) {
      this.props.history.push("/client-login");
    }
  }
  

  render() {
    this.logInCheck(this.props.client);
    const categories = this.displayCategories();
    const name = this.props.client.name || 'client unknown';
    const firstName = name.split(' ')[0];
    return (
      <div className='client-background'>
        <p className='welcome'>One last step {firstName},</p>
        <p className='categories-instructions'>Please select categories that match your problem.</p>
        <div className="categories">
          {categories}
        </div>
        <div className="button-container">
          <button className='previous-button' onClick={()=> this.props.history.push("/problem-body")} >
            PREV
          </button>
          <button className='next-button' onClick={this.handleSubmit} >
            NEXT
          </button>
        </div>
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  client: state.client,
  clientID: state.problem.client,
  title: state.problem.title,
  body: state.problem.body,
  categories: state.categories
});

export const mapDispatchToProps = dispatch => ({
  addCategories: categories => dispatch(addCategories(categories)),
  selectCategories: categories => dispatch(selectCategories(categories))
});

ProblemCategory.propTypes = {
  addCategories: PropTypes.func,
  categories: PropTypes.array,
  client: PropTypes.object,
  selectCategories: PropTypes.func,
  history: PropTypes.object,
  title: PropTypes.string,
  body: PropTypes.string,
  clientID: PropTypes.number
};

export default connect(mapStateToProps, mapDispatchToProps)(ProblemCategory);