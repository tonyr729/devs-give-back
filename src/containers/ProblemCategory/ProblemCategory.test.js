import { ProblemCategory} from './ProblemCategory';
import React from 'react';
import { shallow } from 'enzyme';
jest.mock('../../helpers/DatabaseHelper.js', () => {
  return jest.fn().mockImplementation(() => {
    return {
      pullCategoriesFromDatabase: jest.fn().mockImplementation(() => Promise.resolve({category: 'category'})),
      writeProblemToDatabase: jest.fn().mockImplementation(() => Promise.resolve({problem: 'problem'}))
    };
  });
});


describe('ProblemCategory', () => {
  it('should match snapshot', () => {
    const mockProps = {
      client: {name: 'tony'},
      addCategories: jest.fn()
    };
    const wrapper = shallow(<ProblemCategory {...mockProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  describe('componentDidMount', () => {
    let mockProps;
    let wrapper;

    beforeEach(() => {
      mockProps = {
        client: {name: 'tony'},
        addCategories: jest.fn()
      };
      wrapper = shallow(<ProblemCategory {...mockProps} />);
    });

    it('should call given functions on mount of component', () => {
      expect(wrapper.instance().database.pullCategoriesFromDatabase).toHaveBeenCalled();
      expect(wrapper.instance().props.addCategories).toHaveBeenCalled();
    });
  });

  describe('displayCategories', () => {
    let mockProps;
    let wrapper;

    beforeEach(() => {
      mockProps = {
        client: {name: 'tony'},
        addCategories: jest.fn(),
        categories: ['categoryx', 'categoryx', 'categoryx']
      };
      wrapper = shallow(<ProblemCategory {...mockProps} />);
    });

    it('should render categories if they exist', () => {
      expect(wrapper.find("CategoryButton").length).toEqual(3);
    });
  });

  describe('selectCategories', () => {
    let mockProps;
    let wrapper;

    beforeEach(() => {
      mockProps = {
        client: {name: 'tony'},
        addCategories: jest.fn(),
        categories: ['categoryx', 'categoryx', 'categoryx']
      };
      wrapper = shallow(<ProblemCategory {...mockProps} />);
    });

    it('should add a category to state', () => {
      const expected = 'cat';
      wrapper.instance().selectCategory(expected);
      expect(wrapper.state('categories')).toEqual([expected]);
    });

    it('remove a category from state if it already exists in state', () => {
      const expected = ['cat'];
      wrapper.setState({
        categories: expected
      });

      wrapper.instance().selectCategory(expected[0]);

      expect(wrapper.state('categories')).toEqual([]);
    });
  });

  describe('handleSubmit', () => {
    let mockProps;
    let wrapper;

    beforeEach(() => {
      mockProps = {
        client: {name: 'tony'},
        addCategories: jest.fn(),
        categories: ['categoryx', 'categoryx', 'categoryx'],
        title: 'title',
        body: 'body',
        clientID: 1,
        selectCategories: jest.fn(),
        history: {
          push: jest.fn()
        }
      };
      wrapper = shallow(<ProblemCategory {...mockProps} />);
    });

    it('should call the given functions ', () => {
      wrapper.instance().handleSubmit();

      expect(wrapper.instance().props.selectCategories).toHaveBeenCalled();
      expect(wrapper.instance().database.writeProblemToDatabase).toHaveBeenCalled();
      expect(wrapper.instance().props.history.push).toHaveBeenCalled();
    });
  });

  describe('logInCheck', () => {
    let mockProps;
    let wrapper;

    beforeEach(() => {
      mockProps = {
        client: {},
        addCategories: jest.fn(),
        categories: ['categoryx', 'categoryx', 'categoryx'],
        history: {
          push: jest.fn()
        }
      };
      wrapper = shallow(<ProblemCategory {...mockProps} />);
    });
    
    it('should redirect if no user', () => {
      wrapper.instance().logInCheck(mockProps.client);
      
      expect(wrapper.instance().props.history.push).toHaveBeenCalledWith("/client-login");
    });
  });

});