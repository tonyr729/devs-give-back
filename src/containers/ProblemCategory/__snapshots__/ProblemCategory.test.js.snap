// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProblemCategory should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProblemCategory
    addCategories={[Function]}
    client={
        Object {
            "name": "tony",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p
          className="welcome"
>
          One last step 
          tony
          ,
</p>,
        <p
          className="categories-instructions"
>
          Please select categories that match your problem.
</p>,
        <div
          className="categories"
/>,
        <div
          className="button-container"
>
          <button
                    className="previous-button"
                    onClick={[Function]}
          >
                    <img
                              className="left-arrow-img"
                              src="left_arrow.svg"
                    />
          </button>
          <button
                    className="next-button"
                    onClick={[Function]}
          >
                    <img
                              className="right-arrow-img"
                              src="right_arrow.svg"
                    />
          </button>
</div>,
      ],
      "className": "client-background",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "One last step ",
            "tony",
            ",",
          ],
          "className": "welcome",
        },
        "ref": null,
        "rendered": Array [
          "One last step ",
          "tony",
          ",",
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Please select categories that match your problem.",
          "className": "categories-instructions",
        },
        "ref": null,
        "rendered": "Please select categories that match your problem.",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": null,
          "className": "categories",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="previous-button"
              onClick={[Function]}
>
              <img
                            className="left-arrow-img"
                            src="left_arrow.svg"
              />
</button>,
            <button
              className="next-button"
              onClick={[Function]}
>
              <img
                            className="right-arrow-img"
                            src="right_arrow.svg"
              />
</button>,
          ],
          "className": "button-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <img
                className="left-arrow-img"
                src="left_arrow.svg"
/>,
              "className": "previous-button",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "left-arrow-img",
                "src": "left_arrow.svg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <img
                className="right-arrow-img"
                src="right_arrow.svg"
/>,
              "className": "next-button",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "right-arrow-img",
                "src": "right_arrow.svg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p
            className="welcome"
>
            One last step 
            tony
            ,
</p>,
          <p
            className="categories-instructions"
>
            Please select categories that match your problem.
</p>,
          <div
            className="categories"
/>,
          <div
            className="button-container"
>
            <button
                        className="previous-button"
                        onClick={[Function]}
            >
                        <img
                                    className="left-arrow-img"
                                    src="left_arrow.svg"
                        />
            </button>
            <button
                        className="next-button"
                        onClick={[Function]}
            >
                        <img
                                    className="right-arrow-img"
                                    src="right_arrow.svg"
                        />
            </button>
</div>,
        ],
        "className": "client-background",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "One last step ",
              "tony",
              ",",
            ],
            "className": "welcome",
          },
          "ref": null,
          "rendered": Array [
            "One last step ",
            "tony",
            ",",
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Please select categories that match your problem.",
            "className": "categories-instructions",
          },
          "ref": null,
          "rendered": "Please select categories that match your problem.",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": null,
            "className": "categories",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="previous-button"
                onClick={[Function]}
>
                <img
                                className="left-arrow-img"
                                src="left_arrow.svg"
                />
</button>,
              <button
                className="next-button"
                onClick={[Function]}
>
                <img
                                className="right-arrow-img"
                                src="right_arrow.svg"
                />
</button>,
            ],
            "className": "button-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <img
                  className="left-arrow-img"
                  src="left_arrow.svg"
/>,
                "className": "previous-button",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "left-arrow-img",
                  "src": "left_arrow.svg",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <img
                  className="right-arrow-img"
                  src="right_arrow.svg"
/>,
                "className": "next-button",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "right-arrow-img",
                  "src": "right_arrow.svg",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
